### DISCUSSION TITLE
EMQX 5.0.2 Real client IP

### DISCUSSION THREAD

**Discussion by amnonbb** [emqx/emqx#8409] [Q&A] [answered]
I don't if it's bug or i miss something.
I use nginx to proxy websocket connection, and does not see real clint IP.
With same configuration on emqx 4 it's was not issue.

**Comment by zmstone** [comment]
Hi
I made a script to help test it (using haproxy though):
https://github.com/emqx/emqx/blob/b7053fa2626fe0e49beb8923051490e852115aa8/scripts/start-two-nodes-in-docker.sh
e.g. `start-two-nodes-in-docker.sh emqx/emqx:5.0.2`

It works as expected in my test.

Like 5.0, 4.4, `proxy_protocol` should be enabled for the listener. 
the important part is to make sure:

`listeners.tcp.default.proxy_protocol = true` in `emqx.conf`

or if override from environment variables:

`-e EMQX_listeners__ws__default__proxy_protocol=true`

If this config does not work for you, please share your nginx config?

**Reply by amnonbb** [reply]
For tcp listener it's work.
For ws listener does not.

> location / {
        proxy_pass                          http://msg_cluster/mqtt;
        proxy_http_version                  1.1;
        proxy_set_header Upgrade            $http_upgrade;
        proxy_set_header Connection         "Upgrade";
        proxy_set_header Host               $host;
        proxy_set_header X-Real-IP          $remote_addr;
        proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Port   $server_port;
        proxy_set_header X-Forwarded-Proto  $scheme;
        proxy_read_timeout 310s;
        proxy_buffering off;
    }

**Reply by amnonbb** [reply]
Ah for ws listener you use:
proxy_protocol = true

If i do the same nginx does not work at all

**Reply by zmstone** [reply]
Thanks for the clarification.
I didn’t know you were using the headers.
Will have a closer look later

**Comment by zmstone** [comment]
Hi @amnonbb 

`listeners.ws.default.websocket.proxy_address_header`
`listeners.ws.default.websocket.proxy_port_header`

These two config default values are for some reason changed to all-lowercase letters:
`x-forwarded-for` and `x-forwarded-port`
However, the http server library we use treat the headers case sensitive.
(although it seems, per [RFC7239](https://datatracker.ietf.org/doc/html/rfc7239), they should be case insensitive)

I will submit a PR to revert it back to more conventional values (with uppercase letters).
Before we can cut a new release, you can explicitly set these configs: 
`listeners.ws.default.websocket.proxy_address_header="X-Forwarded-For"`
and
`listeners.ws.default.websocket.proxy_port_header="X-Forwarded-Port"`
as a workaround.

**Comment by zmstone** [comment]
seems to be an unintended change in a big-bang refactoring:
e12a629498436150f41f599d8f4217bd53a72c06

**Reply by zmstone** [reply]
no, sorry, the cases are actually lowered at runtime in 4.3,
now I'm not sure which part went wrong. will try to reproduce it using nginx.

**Comment by zmstone** [comment, marked as answer]
will be fixed in the next release. 5.0.3:
https://github.com/emqx/emqx/pull/8415
